src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -22,281 +20,70 @@ namespace Microsoft.CodeAnalysis
-            _projectIdToTrackerMap = projectIdToTrackerMap;
-            _linkedFilesMap = linkedFilesMap;
-            _dependencyGraph = dependencyGraph;
-            _projectIdToProjectMap = ImmutableHashMap<ProjectId, Project>.Empty;
-            _version = version;
-            _lazyLatestProjectVersion = lazyLatestProjectVersion;
-
-            CheckInvariants();
-        }
 
-        internal Solution(
-            Workspace workspace,
-            SolutionInfo info)
-            : this(
-                workspace.PrimaryBranchId,
-                workspaceVersion: 0,
-                solutionServices: new SolutionServices(workspace),
-                id: info.Id,
-                filePath: info.FilePath,
-                version: info.Version,
-                projectIds: null,
-                idToProjectStateMap: ImmutableDictionary<ProjectId, ProjectState>.Empty,
-                projectIdToTrackerMap: ImmutableDictionary<ProjectId, CompilationTracker>.Empty,
-                linkedFilesMap: ImmutableDictionary.Create<string, ImmutableArray<DocumentId>>(StringComparer.OrdinalIgnoreCase),
-                dependencyGraph: ProjectDependencyGraph.Empty,
-                lazyLatestProjectVersion: null)
+        private Solution(SolutionState state)
         {
-            _lazyLatestProjectVersion = new Lazy<VersionStamp>(() => ComputeLatestProjectVersion());
+            _projectIdToProjectMap = ImmutableHashMap<ProjectId, Project>.Empty;
+            _state = state;
         }
 
-        internal Solution WithNewWorkspace(Workspace workspace, int workspaceVersion)
+        internal Solution(Workspace workspace, SolutionInfo info)
+            : this(new SolutionState(workspace, info))
         {
-            var services = workspace != _solutionServices.Workspace
-                ? new SolutionServices(workspace)
-                : _solutionServices;
-
-            // Note: this will potentially have problems if the workspace services are different, as some services
-            // get locked-in by document states and project states when first constructed.
-            return CreatePrimarySolution(branchId: workspace.PrimaryBranchId, workspaceVersion: workspaceVersion, services: services);
         }
 
-        private VersionStamp ComputeLatestProjectVersion()
-        {
-            // this may produce a version that is out of sync with the actual Document versions.
-            var latestVersion = VersionStamp.Default;
-            foreach (var projectId in this.ProjectIds)
-            {
-                var project = this.GetProject(projectId);
-                latestVersion = project.Version.GetNewerVersion(latestVersion);
-            }
-
-            return latestVersion;
-        }
+        internal SolutionState State => _state;
 
-        internal int WorkspaceVersion
-        {
-            get { return _workspaceVersion; }
-        }
+        internal int WorkspaceVersion => _state.WorkspaceVersion;
```

Comment:
should I remove all internal API from solution and caller to get information through solution.State ?